surefire-provider/src/main/java/org/arquillian/smart/testing/surefire/provider/LoaderVersionExtractor.java

```diff
@@ -0,0 +1,154 @@
+                List<MavenLibrary> wrappedLibrary = Arrays.asList(new MavenLibrary[] {mavenLibrary});
+                Map<MavenLibrary, String> implTitleWithVersion = getTitleWithVersion(wrappedLibrary, loader);
+                loaderWithLibraryVersions.put(loader, implTitleWithVersion);
+            }
+        } else {
+            Map<MavenLibrary, String> implTitleWithVersion = getTitleWithVersion(initLibraries, loader);
+            loaderWithLibraryVersions.put(loader, implTitleWithVersion);
+        }
+
+        return loaderWithLibraryVersions.get(loader).get(mavenLibrary);
+    }
+
+    private static Map<MavenLibrary, String> getTitleWithVersion(List<MavenLibrary> libraries, ClassLoader classLoader) {
+        Map<MavenLibrary, String> implTitleWithVersion = new HashMap<>();
+        ArrayList<MavenLibrary> librariesToFind = new ArrayList<>(libraries);
+        try {
+            Enumeration<URL> manifests = classLoader.getResources("META-INF/MANIFEST.MF");
+
+            while (manifests.hasMoreElements()) {
+                String manifestURL = manifests.nextElement().toString();
+
+                Optional<MavenLibrary> matched =
+                    librariesToFind.parallelStream().filter(library -> manifestURL.matches(library.getRegex())).findFirst();
+
+                if (matched.isPresent()) {
+                    MavenLibrary matchedLibrary = matched.get();
+                    String startWithVersion = manifestURL.replaceAll(matchedLibrary.getLeadingRegex(), "");
+                    String version = startWithVersion.substring(0, startWithVersion.indexOf(File.separator));
+                    implTitleWithVersion.put(matchedLibrary, version);
+                    librariesToFind.remove(matchedLibrary);
+                }
+            }
+        } catch (Exception e) {
+            logger.log(Level.WARNING,
+                "Exception {0} occurred while resolving manifest files",
+                e.getMessage());
+        }
+
+        return implTitleWithVersion;
+    }
+
+    static class MavenLibrary {
+        private final String groupId;
+        private final String artifactId;
+
+        MavenLibrary(String groupId, String artifactId) {
+
+            this.groupId = groupId;
+            this.artifactId = artifactId;
+        }
+
+        String getRegex() {
+            return Pattern.compile(getLeadingString()
+                + ".*"
+                + File.separator
+                + artifactId
+                + "-.*\\.jar.*").pattern();
+        }
+
+        String getLeadingRegex() {
+            return Pattern.compile(getLeadingString()).pattern();
+        }
+
+        private String getLeadingString() {
+            return ".*"
+                + File.separator
+                + groupId.replaceAll("\\.", File.separator)
+                + File.separator
+                + artifactId
+                + File.separator;
+        }
+
+        @Override
+        public boolean equals(Object o) {
```

Comment:
It's autogenerated from IDE and doesn't have `{` `}` for `if`s. Also  last two `if`s are rather cryptic for my taste.